### PR TITLE
Transition to `linopy` implementation

### PR DESCRIPTION
**Pull request by FabianHofmann**
Here we try to make the CI functioning

### CODE REVIEW

**Comment by fneum** [comment]
Yes, I've been making this change back and forth. The problem is that it will not work if you set retrieve_databundle to false...

**Comment by FabianHofmann** [comment]
Ah! I see

**Comment by fneum** [comment]
I'm not sure what a good solution would be.

**Comment by FabianHofmann** [comment]
Neither do I... Do we still need the `retrieve_data_bundle` setting on pypsa-eur? I never use it. 

Alternative but not pretty would be:
* load pypsa-eur config in Snakefile
* make an input function to choose between `pypsaeur` function or direct paths

**Comment by fneum** [comment]
Would definitely keep the retrieve_databundle, because you can directly run snakemake on a cloned repository without manually downloading and extracting.

**Comment by FabianHofmann** [comment]
On a discussion with @euronion we agreed that not relying on pypsaeur default config, makes the world messy. Also we though that disabling the retrieve_databundle config entry in pypsa-eur would be cleaner in the long term. 

In case you have a locally stored bundle that you copied into your freshly cloned pypsa-eur repo, you can simply call 
```bash
snakemake --touch data/bundle/*
```
and it skips the retrieval

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -103,7 +103,5 @@ jobs:
         run: |
           conda activate pypsa-eur
           conda list
-          cp test/config.overnight.yaml config.yaml
-          snakemake -call
-          cp test/config.myopic.yaml config.yaml
-          snakemake -call
+          snakemake -call --configfile test/config.overnight.yaml
+          snakemake -call --configfile test/config.myopic.yaml
```
Comment (fneum) [inline .github/workflows/ci.yaml:107]: I like it, saves a lot of code.

File: config.default.yaml [hunk 2]
```diff
@@ -378,22 +378,23 @@ solving:
...
-    threads: 4
-    method: 2 # barrier
-    crossover: 0
-    BarConvTol: 1.e-6
-    Seed: 123
-    AggFill: 0
-    PreDual: 0
-    GURO_PAR_BARDENSETHRESH: 200
-    #FeasibilityTol: 1.e-6
+    options:
```
Comment (fneum) [inline config.default.yaml:381]: yes, the `.pop()` was always weird. But needs a note in release notes.

**Comment by FabianHofmann** [comment]
Johannes already found the bug, the reduced config.*.yaml files that I made seem to result in  merge conflicts. I'll fix them

File: scripts/solve_network.py [hunk 3]
```diff
@@ -290,6 +290,7 @@ def solve_network(n, config, opts='', **kwargs):
     logging.basicConfig(filename=snakemake.log.python,
                         level=snakemake.config['logging_level'])
 
+    print(snakemake.config)
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply scripts/solve_network.py:293]: indeed :+1:

**Review by fneum — APPROVED** [review]

**Comment by FabianHofmann** [comment]
Because of the nested config update, I had to modularize the solver options to come closer to where we want. I think it's a nice side-effect. 
Last hurdle is that CBC does not want to solve the sec-networks. @fneum did you once manage to solve one with cbc?

**Comment by fneum** [comment]
The CI ran at some point in time.

**Comment by pz-max** [comment]
test Discord webhook

**Comment by FabianHofmann** [comment]
@fneum @lisazeyen I think we can merge this. It comes along with some breaking changes in the `solving` part of  configuration yaml. This was necessary to make the config override via snakemake working. But the structure is now modularized and suits better for extension anyway.

**Review by fneum — COMMENTED** [review]
So far, I have checked everything but the solve_network script. That looks all good!

Since it's a big change to switch the solver interface, could you make a comparison run 3-hourly, 45 nodes, with H2 retrofitting, no gas network, all of Europe and make sure the results are the same? I just want to be sure that we don't introduce a bug.

**Comment by FabianHofmann** [comment]
Sure, I'll set it up :)

**Comment by Irieo** [comment]
> So far, I have checked everything but the solve_network script. That looks all good!
> 
> Since it's a big change to switch the solver interface, could you make a comparison run 3-hourly, 45 nodes, with H2 retrofitting, no gas network, all of Europe and make sure the results are the same? I just want to be sure that we don't introduce a bug.

this branch and the master yield same solution for the test: 37 nodes, 3H sampling, all sectors (T-H-B-I-A). H2 network, H2 retrofitting, no gas network.

**Comment by FabianHofmann** [comment]
Thanks @Irieo for testing! @fneum whenever you feel ready, let's merge it. I would not wait too long so we can add tests for #301

**Comment by fneum** [comment]
Great, what I'll do: Since we have not done a release of PyPSA-Eur-Sec since Oct 2021 and the change of solver interface is a major change, we should do a release before merging this PR. I'll cherry-pick the changes unrelated to the solver interface from this PR.

**Comment by fneum** [comment]
Okay, step 1 is done. Sorry @FabianHofmann, the "co-authored by" was not recognised by Github.

**Comment by FabianHofmann** [comment]
@fneum if no objection, I would merge this

**Pull request merged**